src/pipeline.rs

```diff
@@ -1,16 +1,10 @@
-use super::cas::Cas;
-use super::engine;
-use super::error::PipelineError;
+use crate::engine::{Engine, Reader};
+use crate::error::Result;
 
-pub fn run(
-    reader: &mut dyn engine::Reader,
-    engines: Vec<Box<engine::Engine>>,
-) -> Result<(), PipelineError> {
-    reader.initialize()?;
-    while reader.has_next() {
-        let mut cas = Cas::new();
-        reader.execute(&mut cas)?;
-        for engine in &engines {
+pub fn run<R: Reader>(reader: &mut R, engines: &[Box<dyn Engine>]) -> Result<()> {
```

Comment:
We need the engines to use dynamic dispatch, since we have many different types that all implement Engine stored in one collection. However, the reader can and thus should use static dispatch here.

Reply:
This part I need a bit of help understanding it, what does it mean

fn run<R: Reader> ?

Is it like the box pointer I was using ?

Reply:
[This is a little old but still relevant and can probably explain it better than I can.](https://doc.rust-lang.org/1.0.0-beta/book/static-and-dynamic-dispatch.html) Basically, it means the type gets replaced at compile time with a type that implements Reader. In the way with box pointers (using trait objects) the type isn't used at compile time to replace method calls, but instead a lookup table is used that has the addresses of the methods. Trait objects are very similar to how Java does generics